runtime/collect/StackMapParser.cpp

```diff
@@ -39,26 +39,35 @@ void parseStackMap() {
       void *ip = (void *)(func.FunctionAddress + InstructionOffset);
       int32_t NumDeopts = *(int32_t *)(stackMapRecord + 24 + 2 * 12);
       uint16_t RelocationOffset = 0;
-      for (uint16_t j = 4 + NumDeopts; j < NumLocations; j += 2) {
-        uint8_t type = *(uint8_t *)(stackMapRecord + 16 + j * 12);
-        if (type == 5) {
+      for (uint16_t j = 3 + NumDeopts; j < NumLocations; j += 2) {
+        uint8_t base_type = *(uint8_t *)(stackMapRecord + 16 + j * 12);
+        if (base_type == 5) {
           // a ConstantOffset gc root is one which corresponds to something the
           // compiler was able to statically determine was a constructor with
           // zero children. Such terms do not actually live on the heap and
           // thus do not need to be relocated.
           RelocationOffset++;
           continue;
         }
-        if (type != 3) {
+        if (base_type != 3) {
           abort();
         }
-        int32_t offset = *(int32_t *)(stackMapRecord + 24 + j * 12);
+        uint8_t derived_type = *(uint8_t *)(stackMapRecord + 16 + (j + 1) * 12);
```

Comment:
For example, as above this kind of thing is difficult to grok - a comment mapping the pointer computation to a path into the schema would be good (e.g. for the sake of demonstration on a contrived example not related to this particular line, something like `StkMapRecord[i].Location[j].DwarfRegNum` would clarify exactly what bytes we're expecting to access.

There is a bit of an additional burden involved with keeping this in sync, I guess - how often is this code likely to be modified? If it's only once in a while, then probably worth the effort of keeping the comments.

Reply:
Good call about the comments. I'll go ahead and go through this function and add them.